### ISSUE TITLE
SQL error from Circles App after migration of NC from 18.0.0 to 18.0.1

### ISSUE THREAD

**Issue by Volker-K** [nextcloud/circles#373]
Upgrade of our testing environment from NC 18.0.0 to 18.0.1
Circles app was obviously automatically deactivated during upgrade.

Activation of the app threw an SQL error:

```
An exception occurred while executing 'SELECT `m`.`user_id`, `m`.`user_type`, `m`.`circle_id`, `m`.`level`, `m`.`status`, `m`.`note`, `m`.`contact_id`, `member_id`, `m`.`contact_meta`, `m`.`joined` FROM `oc_circles_members` `m` ORDER BY `m`.`joined` ASC': SQLSTATE[42S22]: Column not found: 1054 Unknown column 'm.contact_id' in 'field list' 
```

Error persists even after deleting and re-installing Circles. 

System here: 
MariaDB 10.2.25
Apache2 2.4
PHP 7.2.5

**Comment by Volker-K** [comment]
After updating the Circles App the error chages:

```
An exception occurred while executing 'INSERT INTO `oc_circles_members` (`circle_id`, `user_id`, `user_type`, `level`, `status`, `joined`, `instance`, `contact_id`) VALUES(?, ?, ?, ?, ?, ?, ?, ?)' with params ["266877b9d4f985", "<mymailaddress>", "3", "1", "Member", "2019-11-22 09:42:57", "", ""]: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '266877b9d4f985-<mymailaddress>-3--' for key 'PRIMARY'
```

**Comment by torbennehmer** [comment]
I can confirm the error of the original post in my installation. I updated OC 17 to 18.0.1. Unfortunately I don't remember the patch level of the old 17 installation.

After upgrade, circles - among other apps - were disabled. Enabling it, completely disabled the installation due to some problem with abstract methods, presumably to an too old circles installation.

After manually updating circles to the latest download from the App Store, the missing column error above surfaced when I try to enable the circles app.

Is there any way I can manually fix this? I have only a few user assignments on this installation, so dropping the circle members and re-adding them afterwards would be easy for me.

**Comment by Volker-K** [comment]
So, I had a look into the oc_circles_members table. Since the row that was duplicate was a circle memeber added via eMail-address I hoped it would only happen to these users.
So I deleted all "user_type=3"-rows from oc_circles_members and tried: Yes, a change, the row that was causing the duplicate key error now was a row with a regular user.
Delete and reinstall the latest circles-version didn't solve the problem.

Is there any chance to get this fixed or just a workaround? Ist the table's definition correct? If so, why does circles try to add rows already existing? Can I do anything to fix it?

**Comment by Volker-K** [comment]
Migrated NC to 18.0.3
Migrated Circles to 0.18.5
Same error:
``Es ist ein Fehler bei der Anfrage aufgetreten. Es kann nicht fortgefahren werden.
An exception occurred while executing 'INSERT INTO `oc_circles_members` (`circle_id`, `user_id`, `user_type`, `level`, `status`, `joined`, `instance`, `contact_id`) VALUES(?, ?, ?, ?, ?, ?, ?, ?)' with params ["266877b9d4f985", "volker.koenig@gmail.com", "3", "1", "Member", "2019-11-22 09:42:57", "", ""]: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '266877b9d4f985-volker.koenig@gmail.com-3--' for key 'PRIMARY'``

Will it ever be noticed as a bug? Any hints how to fix it? @daita? Anyone? 
Please?

**Comment by daita** [comment]
Hello,

I thought this was fixed already, I will have a look into it.

**Comment by Volker-K** [comment]
There's a bunch of issues with Circles and NC18. Right now I can't migrate my installation to NC18 w/o losing all circles users have created. It seems to be a database problem, with PostgreSQL I found an issue that begun with a failed DROP PRIMARY KEY on this table, I get a duplicate key error. Seems to be sort of siblings.
Is there any chance to get the database repaired manually? Any hint?
@daita?

**Comment by Volker-K** [comment]
@torbennehmer  Did you meanwhile find any fix for this problem? Already asked for help in nc forum https://help.nextcloud.com/t/wasnt-able-to-activate-any-version-of-circles-app-after-migrating-to-nc-18/82074

**Comment by torbennehmer** [comment]
@Volker-K Actually, my installation has started to work again, I am not completly sure, what has changed. If I remember correctly, what I did was something along this:
1. Drop the cricles plugin from the installation entirely
2. Upgrade to current NC
3. Redownload the circles plugin again from the app store
It might me releated to this, as I had my installation run for a few days without the plugin. 
Please remember, that I'm guessing here, because at some point in the past, the circles plugin just started to work again. I have never tried to trace it down above that, especially since I tried quite a few things even down to the DB level, trying to isolate the key violation. 
I guess, just dropping the circles data and recreating its settings is no option for you?

Apart from that I'm at a loss, unfortunately, I have never discovered, what actually changed here to make the plugin work again. (Shame on me, as a software developer I hate this kind of riddles...)

Best regards, Torben

**Comment by Volker-K** [comment]
@torbennehmer Thank you for your answer.
I think that my problems are caused by something that went wrong with the first app versions for NC18 and broke the database structure. There's an error reported for PostgreSQL Databases where a primary key couldn't be dropped in the same table I've encountered a missing column first and then, with a later version of the app, primary key validations.
So possibly migrating my testing environment from NC17 to NC18 would work now if I would restore it to NC17 in the first place.
I believe the data model to be quite obvoious, perhaps you could post the output of DESC <tablename> for the oc_circles-tables after migration so I could analyze and understand the errors and prepare migration to NC18 in our live syste?

**Comment by torbennehmer** [comment]
@Volker-K Of course, here it is (albeit from MySQL, not Postgres), my backup script would recreate the oc_circles tables as follows:

```
CREATE TABLE `oc_circles_clouds` (
  `cloud_id` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `address` varchar(128) COLLATE utf8mb4_bin NOT NULL,
  `status` smallint(6) NOT NULL,
  `note` varchar(255) COLLATE utf8mb4_bin DEFAULT NULL,
  `created` datetime DEFAULT NULL,
  PRIMARY KEY (`cloud_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;

CREATE TABLE `oc_circles_groups` (
  `circle_id` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `group_id` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `level` smallint(6) NOT NULL,
  `note` varchar(255) COLLATE utf8mb4_bin DEFAULT NULL,
  `joined` datetime DEFAULT NULL,
  PRIMARY KEY (`circle_id`,`group_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;

CREATE TABLE `oc_circles_links` (
  `id` smallint(5) unsigned NOT NULL AUTO_INCREMENT,
  `status` smallint(6) NOT NULL,
  `circle_id` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `unique_id` varchar(64) COLLATE utf8mb4_bin DEFAULT NULL,
  `address` varchar(128) COLLATE utf8mb4_bin NOT NULL,
  `token` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `key` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `creation` datetime DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;

CREATE TABLE `oc_circles_members` (
  `circle_id` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `user_id` varchar(128) COLLATE utf8mb4_bin NOT NULL,
  `instance` varchar(255) COLLATE utf8mb4_bin NOT NULL DEFAULT '',
  `user_type` smallint(6) NOT NULL DEFAULT 1,
  `contact_id` varchar(127) COLLATE utf8mb4_bin NOT NULL,
  `level` smallint(6) NOT NULL,
  `status` varchar(15) COLLATE utf8mb4_bin DEFAULT NULL,
  `note` varchar(255) COLLATE utf8mb4_bin DEFAULT NULL,
  `joined` datetime DEFAULT NULL,
  `member_id` varchar(15) COLLATE utf8mb4_bin DEFAULT NULL,
  `contact_meta` varchar(1000) COLLATE utf8mb4_bin DEFAULT NULL,
  `contact_checked` smallint(6) DEFAULT NULL,
  PRIMARY KEY (`circle_id`,`user_id`,`user_type`,`contact_id`,`instance`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;

CREATE TABLE `oc_circles_shares` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `unique_id` varchar(32) COLLATE utf8mb4_bin DEFAULT NULL,
  `circle_id` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `source` varchar(15) COLLATE utf8mb4_bin NOT NULL,
  `type` varchar(15) COLLATE utf8mb4_bin NOT NULL,
  `author` varchar(64) COLLATE utf8mb4_bin NOT NULL,
  `cloud_id` varchar(128) COLLATE utf8mb4_bin DEFAULT 'null',
  `headers` varchar(4000) COLLATE utf8mb4_bin DEFAULT NULL,
  `payload` varchar(4000) COLLATE utf8mb4_bin NOT NULL,
  `creation` datetime DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=3 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;

CREATE TABLE `oc_circles_tokens` (
  `circle_id` varchar(31) COLLATE utf8mb4_bin NOT NULL,
  `user_id` varchar(255) COLLATE utf8mb4_bin NOT NULL,
  `share_id` bigint(20) NOT NULL,
  `token` varchar(31) COLLATE utf8mb4_bin NOT NULL,
  `password` varchar(127) COLLATE utf8mb4_bin NOT NULL,
  `member_id` varchar(15) COLLATE utf8mb4_bin DEFAULT NULL,
  `accepted` smallint(6) DEFAULT NULL,
  PRIMARY KEY (`circle_id`,`user_id`,`share_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED;

```

**Comment by Volker-K** [comment]
Thank you, @torbennehmer , unfortunately the tables did get some id-columns that will be filled during migration. I don't have ressources to do this, would have to inspect and uderstand the code first.
@daita no help available with this? I would have to remove Circles completely from our system to get the migration to NC 18 done.

**Comment by daita** [comment]
Do you still have issue with the last version of circles for NC18 (0.18.8) ?

**Comment by Volker-K** [comment]
I suspected that the error was a consequence of a first incorrect migration of the app. So I made a quick clone of the live environment and adjusted the configuration to make sure hostnames, databases etc. were correct.
In the config.php file I noticed settings that were not mine, but related to the circles app.
```
  'app_install_overwrite' =>
  array (
    0 => 'admin_notifications',
    1 => 'announcementcenter',
    2 => 'circles',
  ),

```
I tried the migration to NC18 and got the described error, restored the clone to its old state, deleted the settings and the migration to NC18 ran without any error. 

This morning I successfully migrated the live environment to NC18. 
The settings "app_install_overwrite" are not (yet) documented, but apparently NextCloud itself sets them under certain conditions.

More about ap_install_overwrite: https://github.com/nextcloud/documentation/issues/1865
Seems it already exists for some time but isn't documented yet. The option "--force" of app:activate that seems to let NC write this to config.php isn't documented as well but at least mentioned in the help text of the occ-command.
 
I'm sorry that I couldn't analyze this in depth earlier, but thanks to Covid-19 several of our customers have been improvising home office options since March, with the cloud playing a big role of course.
